types/shares.go

```diff
@@ -184,3 +185,296 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
+
+			// every other namespaceID should be a message
+			default:
+				msgShares = append(msgShares, share)
+			}
+		}
+	}
+
+	// pass the raw share data to their respective parsers
+	txs, err := parseTxs(txsShares)
+	if err != nil {
+		return Data{}, err
+	}
+
+	isrs, err := parseIsrs(isrShares)
+	if err != nil {
+		return Data{}, err
+	}
+
+	evd, err := parseEvd(evdShares)
+	if err != nil {
+		return Data{}, err
+	}
+
+	msgs, err := parseMsgs(msgShares)
+	if err != nil {
+		return Data{}, err
+	}
+
+	return Data{
+		Txs:                    txs,
+		IntermediateStateRoots: isrs,
+		Evidence:               evd,
+		Messages:               msgs,
+	}, nil
+}
+
+// parseTxs collects all of the transactions from the shares provided
+func parseTxs(shares [][]byte) (Txs, error) {
+	// parse the sharse
+	rawTxs, err := processContiguousShares(shares)
+	if err != nil {
+		return nil, err
+	}
+
+	// convert to the Tx type
+	txs := make(Txs, len(rawTxs))
+	for i := 0; i < len(txs); i++ {
+		txs[i] = Tx(rawTxs[i])
+	}
+
+	return txs, nil
+}
+
+// parseIsrs collects all the intermediate state roots from the shares provided
+func parseIsrs(shares [][]byte) (IntermediateStateRoots, error) {
+	rawISRs, err := processContiguousShares(shares)
+	if err != nil {
+		return IntermediateStateRoots{}, err
+	}
+
+	ISRs := make([]tmbytes.HexBytes, len(rawISRs))
+	for i := 0; i < len(ISRs); i++ {
+		ISRs[i] = rawISRs[i]
+	}
+
+	return IntermediateStateRoots{RawRootsList: ISRs}, nil
+}
+
+// parseMsgs collects all evidence from the shares provided
+func parseEvd(shares [][]byte) (EvidenceData, error) {
+	rawEvd, err := processContiguousShares(shares)
+	if err != nil {
+		return EvidenceData{}, err
+	}
+
+	evdList := make(EvidenceList, len(rawEvd))
+
+	// parse into protobuf bytes
+	for i := 0; i < len(rawEvd); i++ {
+		// unmarshal the evidence
+		var protoEvd *tmproto.Evidence
+		err := proto.Unmarshal(rawEvd[i], protoEvd)
+		if err != nil {
+			return EvidenceData{}, err
+		}
+
+		evd, err := EvidenceFromProto(protoEvd)
+		if err != nil {
+			return EvidenceData{}, err
+		}
+
+		evdList[i] = evd
+	}
+
+	return EvidenceData{Evidence: evdList}, nil
+}
+
+// parseMsgs collects all messages from the shares provided
+func parseMsgs(shares [][]byte) (Messages, error) {
+	msgList, err := parseMsgShares(shares)
+	if err != nil {
+		return MessagesEmpty, err
+	}
+
+	return Messages{
+		MessagesList: msgList,
+	}, nil
+}
+
+func processContiguousShares(shares [][]byte) (txs [][]byte, err error) {
+	share := shares[0][NamespaceSize+ShareReservedBytes:]
+	share, txLen, err := parseDelimiter(share)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < len(shares); i++ {
+		var newTxs [][]byte
+		newTxs, share, txLen, err = collectTxFromShare(share, txLen)
+		if err != nil {
+			return nil, err
+		}
+
+		txs = append(txs, newTxs...)
+
+		// if there is no next share
+		if len(shares) <= i+1 {
+			break
+		}
+
+		nextShare := shares[i+1][NamespaceSize+ShareReservedBytes:]
```

Comment:
This is somewhat confusing to read. 

First, `i < len(shares)` because of the for-loop. This means from `len(shares) <= i+1` and `i < len(shares)` always follows `len(shares) == i+1`. So maybe change that break condition to `if len(shares) == i+1 ` (the case `len(shares) < i+1 && i < len(shares)` is impossible).

Then, add a comment on top  of this line: `// At this point we could asset: i+1 < len(shares).` Optionally, you could try to structure the code in a way that is immediately understandable without any code comments (with the clever use of variables).

Reply:
It's definitely difficult to read. I've made some of the changes suggested, but I still think it can be improved more. I struggled with making this more readable the first try too. I'll continue to chew on this while I write the integration test and fix the occasional super small tx size bug. https://github.com/lazyledger/lazyledger-core/pull/261/commits/014a00e0f44e3c7e8c2d9ed8073e1c866c0831e5